### PR TITLE
add NIO event loop as an argument for execute

### PR DESCRIPTION
**Pull request by artemredkin**
Addresses #78

### CODE REVIEW

**Comment by artemredkin** [comment]
@weissi what do you think we need to test? that we use event loop we pass in?

**Comment by weissi** [comment]
> @weissi what do you think we need to test? that we use event loop we pass in?

Two things:
- that it uses the correct event loop
- every API should at least be used by a test case so that there's at least one caller

**Comment by artemredkin** [comment]
@ldewailly done

**Comment by t089** [comment]
I think in its current state, in case of redirect the provided `EventLoop` will be ignored.

In general, is this API really a good idea?

I'm thinking about the future connection pooling. What happens if the pooled connection lives on a different event loop? Should the client then not use this connection and create a new one? Or hop to the event loop of the connection and hop back only in the end? What should happen during the delegate callbacks? On which loop will those be executed?

My understanding was: If you want to tightly control on which event loop the client executes, then initialize it witch a single event loop instead of a event loop group. WDYT?

**Comment by weissi** [comment]
> I'm thinking about the future connection pooling. What happens if the pooled connection lives on a different event loop?

The pool could be keyed on the event loop too but that would as you point out mean that we possibly create more connections than necessary to avoid hopping.

> My understanding was: If you want to tightly control on which event loop the client executes, then initialize it witch a single event loop instead of a event loop group. WDYT?

Good point. That is achievable today through `HTTPClient(eventLoopGroupProvider: .shared(myEventLoop)).execute(...)` (because every `EventLoop` is also an `EventLoopGroup` where `.next()` always returns itself)

I think I'm with @t089 and am unsure if we really need this API or if we should just use a new `HTTPClient` with the EL as the ELG if you want to pin the client to a certain EL. Said that, that also isn't great regarding connection pools because that new `HTTPClient` would then likely have a new connection pool which gets us back to the initial issue that we might create too many connections... @Lukasa thoughts?

**Comment by Lukasa** [comment]
I'm a bit inclined to want to say that the Client should hide this from users.

Creating new clients per event loop seems like an excessive construction. Why not use the event loop as part of the connection key? Users that don't pass an event loop to `execute` are expressing that they don't care (that is, `nil` matches everything), but users that do are requesting a pooled connection on a specific loop. If there isn't one, we just manufacture a new connection for them.

This is strictly no less efficient than creating new clients for every loop.

**Comment by t089** [comment]
> Creating new clients per event loop seems like an excessive construction. Why not use the event loop as part of the connection key? Users that don't pass an event loop to `execute` are expressing that they don't care (that is, `nil` matches everything), but users that do are requesting a pooled connection on a specific loop. If there isn't one, we just manufacture a new connection for them.

This sound doable, although I wonder: if we manufacture a new connection for the explicitly requested event loop, should this connection then also become part of the connection pool? Because in that case, the pool would possibly contain connections that live on event loops that are completely out of the control of the Client. Not sure if this would cause a problem?!

**Comment by Lukasa** [comment]
Yeah, the connection should be part of the pool. Not sure what you mean by an event loop "out of control of the client" though: can you clarify?

**Comment by t089** [comment]
> Yeah, the connection should be part of the pool. Not sure what you mean by an event loop "out of control of the client" though: can you clarify?

Yeah, maybe that is also a non-issue. But I was thinking of this scenario: 
1. Let's say you create the client with `eventLoopGroupProvider: .createNew`.
1. You fire a request with an `EventLoop-#8` from another MTELG.
1. The client will pool the connection with the `EventLoop-#8`
1. You shut down the MTELG.
1. Somebody else fires a request to the same server as above without specifying an event loop.
1. Would the client now try to use a connection that lives on an event loop that has been shut down?

**Comment by Lukasa** [comment]
Yeah, good call, I think the client should confirm that the event loop for the connection is within its event loop group.

**Comment by t089** [comment]
> Yeah, good call, I think the client should confirm that the event loop for the connection is within its event loop group.

So, should connections on "external" event loops become part of the pool or not?

**Comment by Lukasa** [comment]
The client should refuse to make them, to avoid the question being relevant. 😁

**Comment by t089** [comment]
Ahhhh ok, I totally missed that use case: You created the client with a MTELG and you want to make sure that a subsequent request is executed on a specific EL _of this ELG_. Yeah, that's a very good use case. 👍

**Comment by artemredkin** [comment]
is there a way to test if EL is part of our ELG?

**Comment by Lukasa** [comment]
Yes, ELGs expose a property that shows you the loops in them.

**Comment by adtrevor** [comment]
For the case of a connection pool, I wonder what's the cost of hopping to a different event loop (even repeatedly) versus generating a completely new connection to match the event loop requirement, isn't the former typically much less expensive than the latter since network IO is orders of magnitude slower than memory? In this case, wouldn't it have the risk of causing worse performance when the user actually thinks they are doing an optimisation?

Maybe we could do the following with the `eventLoop` argument:
- It _guarantees_ the futures returned by the client are all hopped to `eventLoop`.
- It is used as a hint for the client / connection pool: if a new connection needs to be created, it will use `eventLoop`, otherwise, an existing connection will be used no matter its event loop.

What do you think?

**Comment by Lukasa** [comment]
In general, yes, for a simple request you probably gain nothing by establishing a new TCP connection on your loop instead of reusing one on a different loop. 

However, if you are streaming a very large request or response body, you will need to incur way more coordination overhead, and in those cases it is often worth trying to line up the loops. 

I think your idea is good, but we should consider extending the API to allow users who know they want it to demand a connection on a specific loop.

**Comment by artemredkin** [comment]
@weissi added a test
@Lukasa > I think your idea is good, but we should consider extending the API to allow users who know they want it to demand a connection on a specific loop.
But still from the clients ELG, right?

**Comment by adtrevor** [comment]
@Lukasa  I see, so a finer grained control?

Why not something such as:
```swift
enum EventLoopPreference {
    case indifferent
    case prefers(EventLoop)
    case requires(EventLoop)
}
```

- `.indifferent` is the same as `nil` in the current PR.
- `.prefers(EventLoop)` guarantees futures callbacks are executed on `EventLoop` by hopping them, but imposes no further requirement on connection providers implementations, the goal is to enable them to make the right optimizations without over constraining them.
- `requires(EventLoop)` guarantees futures callbacks are executed on `EventLoop` by ensuring everything is executed on `EventLoop`.

`.prefers(...)` would be destined to cases where the user knows using a specific event loop can offer performance benefits, but also knows that forcing the creation of a new connection would probably outweigh the performance gains. On the other hand, `requires(...)` forces the creation of new connections, better suited for cases such as the long lived streaming @Lukasa describes.

The current implementation would likely treat `.prefers(...)` and `.requires(...)` the exact same way, because it creates a new connection each time, but I'm interested to see this already integrated into the API as I'm not sure of how this could be integrated after the `1.0.0` release without being a breaking change?

**Comment by weissi** [comment]
@adtrevor This might be the right idea. I'm just a little worried that right now we're probably not able to express all the things we want to express in the future and the `enum` is set once we release 1.0.0. What I would probably recommend (is something like):

```
public struct EventLoopPreferrence {
    private enum Preferrence {
        case .indifferent
        case .prefers(EventLoop)
        /* this is private, we can extend in the future */
    }
    private var preferrence: Preferrence

    public static let indifferent = EventLoopPreferrence(preferrence: .indifferent)
    public static func prefers(_ eventLoop: EventLoop) = EventLoopPreferrence(preferrence: eventLoop)
}

extension EventLoopPreferrence: NilLiteralConvertible {
   /* make `nil` equal to `.indifferent` */
}
```

if we do this we support the following on day 1: `nil` (or `.indifferent`) and `.prefers(myEventLoop)` and we don't need to implement `.requires(...)` yet. We can (in a SemVer minor) add any other thing later such as (bad example) `.requiresEitherOr(thisEL, thatEL)` without breaking the public API.

Obviously if we think `.requires` is more useful to start with than `.prefers`, we can also start with that.

**Comment by Lukasa** [comment]
Yeah, I think a finer-grained preference control is good. I like that idea, and think if we go with it we should adopt @weissi's suggestion. I also agree with @artemredkin's reminder that we should only allow users to ask for an event loop from the group the client is using.

**Comment by adtrevor** [comment]
@weissi Thank you, I already saw this pattern in NIO and I like it.
I also agree starting with `.prefers` because `.requires` may cause noticeable performance losses if used incorrectly with a pool IMO.
I'm a little worried about users not realising `.prefers` still guarantees the futures are already hopped to the loop they specified? Maybe there could be a better name? The docs might also be enough since `EventLoopPreferrence` is already a quite "advanced" option one shouldn't use without understanding what it precisely does?

**Comment by artemredkin** [comment]
@weissi @adtrevor great idea, added preferrence

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 1]
```diff
@@ -312,6 +342,23 @@ public class HTTPClient {
...
+            case prefers(EventLoop)
+        }
+
+        var preferrence: Preferrence
+
+        /// Event Loop will be selected by the library.
+        public static let indifferent = EventLoopPreferrence(preferrence: .indifferent)
+        /// Library will try to use provided event loop if possible.
+        public static func prefers(_ eventLoop: EventLoop) -> EventLoopPreferrence { EventLoopPreferrence(preferrence: .prefers(eventLoop)) }
+    }
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClient.swift]: Preference only has one 'r'.

Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: oops, fixed, thanks!

**Comment by t089** [comment]
> I also agree with @artemredkin's reminder that we should only allow users to ask for an event loop from the group the client is using.

This part is still missing, isn't it? Should we add a test for it?

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 2]
```diff
@@ -312,6 +342,25 @@ public class HTTPClient {
...
+            case prefers(EventLoop)
+        }
+
+        var preference: Preference
+
+        /// Event Loop will be selected by the library.
+        public static let indifferent = EventLoopPreference(preference: .indifferent)
+        /// Library will try to use provided event loop if possible.
+        public static func prefers(_ eventLoop: EventLoop) -> EventLoopPreference {
+            return EventLoopPreference(preference: .prefers(eventLoop))
```
Comment (weissi) [inline Sources/AsyncHTTPClient/HTTPClient.swift]: the parameter shouldn't have a name here, "preference" is duplicated. Should be `EventLoopPreference(.prefers(eventLoop))`

**Comment by adtrevor** [comment]
@artemredkin Thanks, that's great!

File: Sources/AsyncHTTPClient/HTTPClient.swift (hunk 2 above)
Comment (artemredkin) [reply Sources/AsyncHTTPClient/HTTPClient.swift]: good idea, fixed!

**Comment by artemredkin** [comment]
@t089 good catch, thank you! updated the PR

**Comment by t089** [comment]
Note, that redirects will ignore the preference parameter, won't they?

https://github.com/swift-server/async-http-client/blob/8fa354df7ccf7139afe4cb9211f17dbcf2c57ed8/Sources/AsyncHTTPClient/HTTPClient.swift#L218

**Comment by t089** [comment]
Also in the current implementation `.prefers(EventLoop)` does what @adtrevor described as `.requires(EventLoop)`: It creates a connection on the requested `EventLoop`. So shouldn't this be called `.requires(EventLoop)`, then?

**Comment by adtrevor** [comment]
@t089 The issue if we use `.requires(...)` is that when the lib evolves to include pooling, there might be big performance losses for people who still have `.requires(...)` option set. That's because it imposes big constraints causing the spin of sometimes useless new connections.

**Comment by t089** [comment]
Ah got it, thanks!

**Comment by adtrevor** [comment]
> Note, that redirects will ignore the preference parameter, won't they?
> 
> https://github.com/swift-server/async-http-client/blob/8fa354df7ccf7139afe4cb9211f17dbcf2c57ed8/Sources/AsyncHTTPClient/HTTPClient.swift#L218

Indeed looks like this part is missing?

**Comment by weissi** [comment]
> @t089
> 
> > Note, that redirects will ignore the preference parameter, won't they?
> > https://github.com/swift-server/async-http-client/blob/8fa354df7ccf7139afe4cb9211f17dbcf2c57ed8/Sources/AsyncHTTPClient/HTTPClient.swift#L218
> 
> Indeed looks like this part is missing?

Sorry, I merged because I didn't see this: Created issue https://github.com/swift-server/async-http-client/issues/88

**Pull request merged**